server/src/main/scala/services/graphstages/ArrivalSplitsGraphStage.scala

```diff
@@ -0,0 +1,307 @@
+      override def onPush(): Unit = {
+        log.debug(s"inSplitsPredictions onPush called")
+        val predictions = grab(inSplitsPredictions)
+
+        log.info(s"Grabbed ${predictions.length} predictions")
+        val updatedFlights = addPredictions(predictions, flightsByFlightId)
+        arrivalsWithSplitsDiff = updatedFlights.values.toSet -- flightsByFlightId.values.toSet
+        flightsByFlightId = updatedFlights
+
+        pushStateIfReady()
+        pullAll()
+      }
+    })
+
+    def pullAll(): Unit = {
+      List(inManifests, inArrivalsDiff, inSplitsPredictions).foreach(in => if (!hasBeenPulled(in)) {
+        log.info(s"Pulling ${in.toString}")
+        pull(in)
+      })
+    }
+
+    def addPredictions(predictions: Seq[(Arrival, Option[ApiSplits])], flightsById: Map[Int, ApiFlightWithSplits]): Map[Int, ApiFlightWithSplits] = {
+      predictions
+        .collect {
+          case (arrival, Some(splits)) => (arrival, splits)
+        }
+        .foldLeft(flightsById) {
+          case (existingFlightsByFlightId, (arrivalForPrediction, predictedSplits)) =>
+            existingFlightsByFlightId.find {
+              case (_, ApiFlightWithSplits(existingArrival, _, _)) => existingArrival.uniqueId == arrivalForPrediction.uniqueId
+            } match {
+              case Some((id, existingFlightWithSplits)) =>
+                val newSplitsSet: Set[ApiSplits] = updateSplitsSet(arrivalForPrediction, predictedSplits, existingFlightWithSplits)
+                existingFlightsByFlightId.updated(id, existingFlightWithSplits.copy(splits = newSplitsSet))
+              case None =>
+                existingFlightsByFlightId
+            }
+        }
+    }
+
+    def updateSplitsSet(arrivalForPrediction: Arrival, predictedSplits: ApiSplits, existingFlightWithSplits: ApiFlightWithSplits): Set[ApiSplits] = {
+      val predictedSplitsWithPaxNumbers = predictedSplits.copy(splits = predictedSplits.splits.map(s => s.copy(paxCount = s.paxCount * ArrivalHelper.bestPax(arrivalForPrediction))), splitStyle = PaxNumbers)
+      val predictedWithEgatesAndFt = predictedSplitsWithPaxNumbers.copy(splits = splitsCalculator.addEgatesAndFastTrack(arrivalForPrediction, predictedSplitsWithPaxNumbers.splits))
+      val newSplitsSet = existingFlightWithSplits.splits.filterNot {
+        case ApiSplits(_, SplitSources.PredictedSplitsWithHistoricalEGateAndFTPercentages, _, _) => true
+        case _ => false
+      } + predictedWithEgatesAndFt
+      newSplitsSet
+    }
+
+    def updateFlightsFromIncoming(arrivalsDiff: ArrivalsDiff, existingFlightsById: Map[Int, ApiFlightWithSplits]): Map[Int, ApiFlightWithSplits] = {
+      log.info(s"${arrivalsDiff.toUpdate.size} diff updates, ${existingFlightsById.size} existing flights")
+      val afterRemovals = existingFlightsById.filterNot {
+        case (id, _) => arrivalsDiff.toRemove.contains(id)
+      }
+      val latestCrunchMinute = getLocalLastMidnight(now().addDays(maxDaysToCrunch))
```

Comment:
should latestCrunchMinute be lastMinuteToCrunch rather?

Reply:
yes good suggestion. i couldn't think of a better name at the time but realised it didn't quite sound right